import os, re
import glob
import shutil
import pandas as pd

##############################
# CONFIG
# can be overwritten by using --configfile <path to config> when calling snakemake
# configfile:"config/config.yaml"

include:
    "rules/init.smk"

NUMBERS = ["01", "02", "03",
    "04",  "05", "06",
    "07", "08", "09",
    "10", "11", "12"]

rule assembly:
    input:
        os.path.join(RESULTS_DIR, "rawReads/concatenated_reads.fastq.gz"),
        os.path.join(RESULTS_DIR, "filteredReads/filtered_reads.fastq"),
        expand(os.path.join(RESULTS_DIR, "subsampledReads/sample_{number}.fastq"), number = NUMBERS),
        expand(os.path.join(RESULTS_DIR, "flye/sample_{number}/assembly_graph.gfa"), number = ["01", "04", "07", "10"]),
        expand(os.path.join(RESULTS_DIR, "canu/sample_{number}"), number = ["02", "05", "08", "11"]),
        expand(os.path.join(RESULTS_DIR, "raven/sample_{number}.gfa"), number = ["03", "06", "09", "12"])
    # output:
    #     touch("status/preprocessing.done")

rule concatenate_reads:
	input:
		expand(os.path.join(READS_DIR, "{reads}.fastq.gz"), reads = READS)
	output:
		os.path.join(RESULTS_DIR, "rawReads/concatenated_reads.fastq.gz")
	shell:
		"cat {input} > {output}"

rule filter:
    input:
        os.path.join(RESULTS_DIR, "rawReads/concatenated_reads.fastq.gz")
    output:
        os.path.join(RESULTS_DIR, "filteredReads/filtered_reads.fastq")
    conda:
        os.path.join(ENV_DIR, "filtlong.yaml")
    shell:
        "filtlong --min_length 2000 --keep_percent 90 {input} > {output}"

rule subsample:
    input:
        os.path.join(RESULTS_DIR, "filteredReads/filtered_reads.fastq")
    output:
        os.path.join(RESULTS_DIR, "subsampledReads/sample_{number}.fastq")
    conda:
        os.path.join(ENV_DIR, "trycycler.yaml")
    params:
        config["subsample"]["genome_size"]
    shell:
        "trycycler subsample --reads {input} --out_dir $(dirname {output}) --genome_size {params}"

rule flye:
    input:
        os.path.join(RESULTS_DIR, "subsampledReads/sample_{number}.fastq")
    output:
        gfa = os.path.join(RESULTS_DIR, "flye/sample_{number}/assembly_graph.gfa"),
        fasta = os.path.join(RESULTS_DIR, "flye/sample_{number}/assembly.fasta")
    conda:
        os.path.join(ENV_DIR, "flye.yaml")
    threads: 8
    shell:
        "flye --nano-hq {input} --out-dir $(dirname {output.gfa}) --threads {threads}"

rule canu:
    input:
        os.path.join(RESULTS_DIR, "subsampledReads/sample_{number}.fastq")
    output:
        directory(os.path.join(RESULTS_DIR, "canu/sample_{number}"))
    params:
        config["subsample"]["genome_size"]
    conda:
        os.path.join(ENV_DIR, "canu.yaml")
    threads: 8
    shell:
        "canu -p canu -d {output} genomeSize={params} -nanopore {input} -maxThreads={threads}"

rule raven:
    input:
        os.path.join(RESULTS_DIR, "subsampledReads/sample_{number}.fastq")
    output:
        gfa=os.path.join(RESULTS_DIR, "raven/sample_{number}.gfa"),
        fasta=os.path.join(RESULTS_DIR, "raven/sample_{number}.fasta")
    conda:
        os.path.join(ENV_DIR, "raven.yaml")
    threads: 8
    shell:
        "raven -F {output.gfa} --output {output.gfa} --threads 8 --disable-checkpoints {input} > {output.fasta}"